src/scf/newton.jl

```diff
@@ -0,0 +1,489 @@
+    Nk = length(basis.kpoints)
+    Ωδφ = similar(φ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Hk = H.blocks[ik]
+        λk = φk' * (Hk*φk)
+        Ωδφk = similar(δφk)
+
+        N1 = size(δφk, 2)
+        N2 = size(φk, 2)
+        @assert N1 == N2
+        N = N1
+
+        for i = 1:N
+            Hδφki = Hk * δφk[:, i]
+            Ωδφk[:, i] = Hδφki
+            for j = 1:N
+                Ωδφk[:, i] -= (φk[:, j]'Hδφki) * φk[:, j]
+                Ωδφk[:, i] -= λk[j, i] * δφk[:, j]
+            end
+        end
+        Ωδφ[ik] = Ωδφk
+    end
+    # ensure proper projection onto the tangent space
+    proj_tangent(Ωδφ, φ)
+end
+
+"""
+    apply_K(basis::PlaneWaveBasis, δφ, φ, ρ, occ)
+
+Compute the application of K to an element in the space tangent to φ
+with the same notations as for the computation of Ω, we have
+      Kδφi = Π(dV * φi),
+Π being the projection on the space tangent to the φ and
+      dV = kernel(δρ)
+where δρ = Σi φi*conj(δφi) + hc.
+"""
+@views function apply_K(basis::PlaneWaveBasis, δφ, φ, ρ, occ)
+    Nk = length(basis.kpoints)
+    n_spin = basis.model.n_spin_components
+
+    ## compute δρ = Σi φi*conj(δφi)
+    δρ_fourier = zeros(complex(eltype(ρ)), size(ρ)...)
+    for (ik, kpt) in enumerate(basis.kpoints)
+        δρk = zeros(eltype(ρ), basis.fft_size)
+        for i = 1:size(φ[ik], 2)
+            φki_real = G_to_r(basis, kpt, φ[ik][:, i])
+            δφki_real = G_to_r(basis, kpt, δφ[ik][:, i])
+            δρk .+= occ[ik][i] .* (conj.(φki_real) .* δφki_real +
+                                   conj.(δφki_real) .* φki_real)
+        end
+        # Check sanity in the density, we should have ∫δρ = 0
+        T = real(eltype(δρk))
+        sum_δρ = sum(δρk)
+        if abs(sum_δρ) > sqrt(eps(T))
+            @warn("Mismatch in δρ", sum_δρ=sum_δρ)
+        end
+        δρk_fourier = r_to_G(basis, complex(δρk))
+        lowpass_for_symmetry!(δρk_fourier, basis)
+        accumulate_over_symmetries!(δρ_fourier[:, :, :, kpt.spin], δρk_fourier,
+                                    basis, basis.ksymops[ik])
+    end
+    mpi_sum!(δρ_fourier, basis.comm_kpts)
+    count = sum(length(basis.ksymops[ik]) for ik in 1:length(basis.kpoints)) ÷ n_spin
+    count = mpi_sum(count, basis.comm_kpts)
+    δρ = G_to_r(basis, δρ_fourier) ./ count
+
+    ## dV = kernel(δρ)
+    dV = apply_kernel(basis, δρ; ρ=ρ)
+
+    ## Kδφi = Π(dV * φi)
+    Kδφ = similar(φ)
+    for ik = 1:Nk
+        kpt = basis.kpoints[ik]
+        φk = φ[ik]
+        dVφk = similar(φk)
+
+        for i = 1:size(φk, 2)
+            φk_r = G_to_r(basis, kpt, φk[:,i])
+            dVφk_r = dV[:, :, :, kpt.spin] .* φk_r
+            dVφk[:, i] = r_to_G(basis, kpt, dVφk_r)
+        end
+        Kδφ[ik] = dVφk
+    end
+    # ensure proper projection onto the tangent space
+    proj_tangent(Kδφ, φ)
+end
+
+#
+# Callback and convergence
+#
+
+"""
+Flag convergence as soon as total energy change drops below tolerance
```

Comment:
As newton is mostly used to get high precision results and energy changes can be eps even if the solution is only converged to sqrt(eps), it seems weird to use this as a termination criterion